Change history time to be equal to the middle of the time bounds

slevis-lmwg opened the pull request:
### Description of changes
This PR subsets the scope of issue #1059 and PR #2445 as a result of the October 2024 conversation in #2445.
This PR changes history time to be equal to the middle of the time bounds. 
This PR does not put instantaneous fields on their own separate history files.

I will also bring submodule changes from https://github.com/ESCOMP/MOSART/pull/106 (was https://github.com/ESCOMP/MOSART/pull/69) and https://github.com/ESCOMP/RTM/pull/39.

### Specific notes

Contributors other than yourself, if any:

Are answers expected to change (and if so in what way)?
No.

Does this create a need to change or add documentation? Did you do so?
Maybe. No.

Testing performed, if any:
Plan to run aux_clm, mosart, rtm test-suites.

slevis-lmwg commented:
I submitted this manual test to confirm that the committed modifications work as intended:
`./create_test SMS_Lm1.f10_f10_mg37.I1850Clm60BgcCropCmip6waccm.derecho_gnu.clm-basic -c /glade/campaign/cgd/tss/ctsm_baselines/ctsm5.3.009`
Check on Friday.

slevis-lmwg commented:
The previous test completed its 1 month and the monthly output looked good, but there were annual history files that I could not tell. So I started another test (default is Ly1, but I changed to Ly2) and I added  hist_avgflag_pertape(6) = 'I' to see what happens:
`SMS_Ly2_Mmpi-serial.1x1_brazil.IHistClm60BgcQianRs.derecho_intel.clm-output_bgc_highfreq`

slevis-lmwg commented:
I updated the submodules to point to https://github.com/ESCOMP/MOSART/pull/106 and https://github.com/ESCOMP/RTM/pull/39 and submitted the three corresponding test-suites:
```
./run_sys_tests -s mosart -c mosart1.1.02-ctsm5.2.029 --skip-generate
./run_sys_tests -s rtm -c rtm1_0_80-ctsm5.2.029 --skip-generate
./run_sys_tests -s aux_clm -c ctsm5.3.009 --skip-generate
```

slevis-lmwg commented:
@ekluzek and I agreed on the order that the "hist" PRs would get merged. The order as shown in [Upcoming Tags ](https://github.com/orgs/ESCOMP/projects/46/views/1)is 
#2838 
#2084 
#2052
...and we will follow the same order in mosart/rtm. Before I merge the mosart/rtm "hist" PRs, @ekluzek will merge the work in the "simple bfb" mosart/rtm cards.

slevis-lmwg commented:
@ekluzek review and approval of this PR should take 5 minutes, as it looks the same as the corresponding RTM and MOSART PRs that you reviewed/approved. Thanks :-)

slevis-lmwg commented:
TODOs left for me:
- Update to ctsm5.3.012
- Update .gitmodules
- Run aux_clm, mosart, rtm
- Merge and make tag

slevis-lmwg commented:
derecho testing IN PROGRESS
./run_sys_tests -s aux_clm -c ctsm5.3.012 -g ctsm5.3.013
./run_sys_tests -s rtm -c rtm1_0_82-ctsm5.3.009 -g rtm1_0_82-ctsm5.3.013
./run_sys_tests -s mosart -c mosart1.1.04-ctsm5.3.009 -g mosart1.1.04-ctsm5.3.013

izumi testing

slevis-lmwg commented:
The diffs of the two tests above seem potentially related to the earlier update to mosart1.1.02 (https://github.com/ESCOMP/MOSART/pull/94), but Adrianna's test passed just fine pointing to mosart1.1.02. So I will try the two tests pointing to mosart1.1.02. If that passes, then mosart1.1.03.

slevis-lmwg commented:
I brainstormed for a bit with @billsacks and Bill pointed out/suggested:
- these two are the only long tests with active cism
- the changes appear in the coupler due to the changes in mosart and not due to changes in ctsm
- that he would not expect these diffs (as I also didn't), so he would recommend going through a methodical way of testing, making baselines, and updating the code and the baselines to confirm whether I still get these diffs

slevis-lmwg commented:
@billsacks I mentioned to you a vague memory I had of an issue that could relate to these diffs, and it is this one:
#2542

billsacks commented:
> I mentioned to you a vague memory I had of an issue that could relate to these diffs, and it is this one:
> #2542

Ah, yes. But I don't think that should be the issue with these tests, right?

slevis-lmwg commented:
Right, I don't think so. I think I have now found that the problem starts with the introduction of mosart1.1.03. I see this in a new test today and in my testing from yesterday (somehow I missed the sign when I looked originally). First I will confirm beyond doubt and then I will try bisecting mosart1.1.03 to find the culprit.

samsrabin commented on src/main/histFileMod.F90:
```diff
@@ -3397,18 +3422,23 @@ subroutine htape_timeconst(t, mode)
        dim2id(1) = hist_interval_dimid;  dim2id(2) = time_dimid
-       call ncd_defvar(nfid(t), 'time_bounds', ncd_double, 2, dim2id, varid, &
-          long_name = 'history time interval endpoints')
+       if (tape(t)%hlist(1)%avgflag /= 'I') then  ! NOT instantaneous fields tape
```
Might be prudent to introduce a variable like `is_inst_tape` instead of repeating this conditional everywhere. That way, once you start segregating instantaneous/non variables onto their own tapes, you only have to change it in one place.

slevis-lmwg replied:
Thank you @samsrabin 
I linked your comment to #2445 and will address it no later than at that time.

slevis-lmwg replied:
The alternate if-statement [discussed below ](https://github.com/ESCOMP/CTSM/pull/2838#issuecomment-2492286937)resolves this.

In the context of treating instantaneous tapes differently with this PR and #2445, @samsrabin asked whether we loop through all of a tape's fields to identify instantaneous tapes at runtime. The answer is no, and we agreed that #2445 will handle this by separating all instantaneous fields onto separate tapes and, therefore, no need to worry about doing this now.

slevis-lmwg commented:
> then I will try bisecting mosart1.1.03 to find the culprit.

Looking at https://github.com/ESCOMP/MOSART/pull/70, I have converged on two commits:
We removed two lines: 77494596d594df6c5843747944abf1612c8d5494
Instead of removing the two lines, we added if-statemt that is .false.: 692d1839489d6979c1a241e0f638189186405cfc

I have confirmed that the two lines that we removed caused the diffs.
@ekluzek I will check with you how to resolve this.

slevis-lmwg commented:
Trying a case with cism NOT active and the if-statement still commented out (as in my last test)
`./create_test SMS_D.f10_f10_mg37.I2000Clm60Bgc.derecho_intel -c /glade/campaign/cgd/tss/ctsm_baselines/ctsm5.3.012`

slevis-lmwg commented:
Rerunning this test to generate a baseline, but stuck in the SHAREDLIB_BUILD phase, so I will kill it and try again next week. Besides, I will need to generate a ctsm5.3.014 baseline, so rerunning right now is redundant:
`./create_test RXCROPMATURITYSKIPGEN_Ld1097.f10_f10_mg37.IHistClm60BgcCrop.derecho_intel.clm-cropMonthOutput -c /glade/campaign/cgd/tss/ctsm_baselines/ctsm5.3.012 -g /glade/campaign/cgd/tss/ctsm_baselines/ctsm5.3.013_slevis_ok`

samsrabin commented:
I'm wondering if the "treat a file as instantaneous if its first variable is" might be premature to bring in here instead of #2445. Specifically, I think the `time_bounds` variable should for now be saved no matter what. Not having it messed up the RXCROPMATURITY test, and though I was able to work around it, others might not be.

slevis-lmwg commented:
> [...] Specifically, I think the `time_bounds` variable should for now be saved no matter what. Not having it messed up the RXCROPMATURITY test, and though I was able to work around it, others might not be.

@olyson what do you think about @samsrabin's comment? Most concerning to me would be any vulnerability in the land diagnostic package.

olyson commented:
I don't think the land diagnostics package uses time_bounds.  But I may not understand the issue here.  Seems like we could do a short simulation using this branch once stable and see if there are any problems?  Also should check ILAMB.  So maybe an I2000 case for a test.

samsrabin commented:
And actually, this comment goes for the "exact middle" vs. "end of" change as well. It seems arbitrary (and against the "principle of least astonishment") that the first variable in the `hist_fincl` list should affect this. My understanding was that this would be changed to "exact middle" for all history files for now, and we would just accept that being wrong for instantaneous variables.

samsrabin commented:
A bonus from what I'm proposing: Always (a) including `time_bounds` and (b) setting `time` to the exact middle means that people always have what they need for postprocessing either instantaneous variables (just look at the second value in `time_bounds`) or averaged/etc. variables (either both values in `time_bounds` or the value in `time`).

ekluzek commented:
time_bounds is an expected part of the CF convention, so I do endorse using it for anything with time. That's likely why some tools might expect it to be there.

For instantaneous it should likely be the time bounds of the time-step that was output. So the previous time-step time first to the ending time-step time for the endpoint. You could have both being the same ending time-step, but that doesn't show that it is a model with a finite time-step.

Here's information on the CF Convention attributes. Look up "bounds"...

https://cfconventions.org/Data/cf-conventions/cf-conventions-1.11/cf-conventions.html#attribute-appendix

samsrabin commented:
@ekluzek That link contains the following text, which implies that instantaneous files should actually _not_ have `time_bounds`:

> It is often the case that data values are not representative of single points in time and/or space, but rather of intervals or multidimensional cells. This convention defines a bounds attribute to specify the extent of intervals or cells.

ekluzek commented:
Good point @samsrabin. That convinces me we should remove it for I fields then. That would then be in line with the convention.

We don't similarly report on the grid cell bounds either (which could be done for 2D grids, but would be harder for unstructured grids), so we shouldn't for Instantaneous time fields either.

billsacks commented:
Earlier discussion led to the conclusion that we should *not* have time_bounds in instantaneous files (though CAM's mistakenly still does): https://github.com/ESCOMP/CAM/issues/1166

slevis-lmwg commented:
I had a quick meeting with @samsrabin 40 minutes ago:
I proposed and he agreed to an alternate version of the if-statement that rightly concerned him. The alternate version eliminates the risk of wrongly labeling a tape "instantaneous" just because the first field is instantaneous.

Oh, good, looks as though we're removing time_bounds from instantaneous tapes, as originally planned :-)

Ok, so I'm testing the alternate if-statement with aux_clm right now and then I will push it to the PR.

slevis-lmwg commented:
@samsrabin back to this test
`/glade/derecho/scratch/slevis/tests_1121-124652de/RXCROPMATURITYSKIPGEN_Ld1097.f10_f10_mg37.IHistClm60BgcCrop.derecho_intel.clm-cropMonthOutput.GC.1121-124652de_int.gddgen`
I get a new error. The problem and solution are not obvious to me from a quick look, but I'm happy to look together if you want.

slevis-lmwg commented:
@samsrabin I'm thinking please hold regarding my last post for now, because in my aux_clm tests with the two subsequent PRs, which included this one, the test passed.

slevis-lmwg commented:
Note to @ekluzek 
- If you're up for approving this one, then I would be ready to move forward with the merge when it's time.
- You have seen it before; it should look very familiar, other than an addition from Sam Rabin pertaining to cropcal_module.py.

Note to self:
Wait for Erik's upcoming mosart/rtm tags before proceeding with this merge.

slevis-lmwg commented:
Updated to rtm1_0_84 and mosart1.1.06 and started on derecho:
```
PASS ./build-namelist_test.pl
PASS lint and black
PASS ./run_ctsm_py_tests -u and -s
OK ./run_sys_tests -s rtm -c rtm1_0_84-ctsm5.3.016 -g rtm1_0_84-ctsm5.3.018
IN PROGRESS ./run_sys_tests -s aux_clm -c ctsm5.3.017 -g ctsm5.3.018
WAIT ./run_sys_tests -s mosart -c mosart1.1.07-ctsm5.3.017 -g mosart.1.1.07-ctsm5.3.018
```
Mosart test-suite waiting for 07 tag to come in with this PR: https://github.com/ESCOMP/MOSART/pull/112

**Notes**
How I check for aux_clm results in tests_0109-170044de:
```
./cs.status.fails | grep -v '17: DI'
grep NORM */*cprnc.out | grep -v time
```
- The first command confirms that all FAILs are baseline diffs, as expected.
- The second line confirms that all baseline diffs are in the time variable, as expected.

ekluzek reviewed (approved):
@slevis-lmwg and I went over this. I had some comments and questions that we worked through. The testing is done so this can be made shortly.

The one suggestion I had was to point out in the ChangeLog that the mosart tag used for the Izumi baseline included the bug-fix that will be made in mosart1.1.07 that will come in the next tag ctsm5.3.019.

Since, testing and baselines had been done, and there are things that have to be done by hand on Izumi we didn't want to have to redo any of the testing.
